lib/Dialect/Arc/Transforms/AllocateState.cpp

```diff
@@ -0,0 +1,166 @@
+//===- AllocateState.cpp --------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "arc-allocate-state"
+
+using namespace mlir;
+using namespace circt;
+using namespace arc;
+
+using llvm::SmallMapVector;
+
+//===----------------------------------------------------------------------===//
+// Pass Implementation
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct AllocateStatePass : public AllocateStateBase<AllocateStatePass> {
+  void runOnOperation() override;
+  void allocateBlock(Block *block);
+  void allocateOps(Value storage, Block *block, ArrayRef<Operation *> ops);
+};
+} // namespace
+
+void AllocateStatePass::runOnOperation() {
+  ModelOp modelOp = getOperation();
+  LLVM_DEBUG(llvm::dbgs() << "Allocating state in `" << modelOp.getName()
+                          << "`\n");
+
+  // Walk the blocks from innermost to outermost and group all state allocations
+  // in that block in one larger allocation.
+  modelOp.walk([&](Block *block) { allocateBlock(block); });
+}
+
+void AllocateStatePass::allocateBlock(Block *block) {
+  SmallMapVector<Value, std::vector<Operation *>, 1> opsByStorage;
+
+  // Group operations by their storage. There is generally just one storage,
+  // passed into the model as a block argument.
+  for (auto &op : *block) {
+    if (isa<AllocStateOp, RootInputOp, RootOutputOp, AllocMemoryOp,
+            AllocStorageOp>(&op))
+      opsByStorage[op.getOperand(0)].push_back(&op);
+  }
+  LLVM_DEBUG(llvm::dbgs() << "- Visiting block in "
+                          << block->getParentOp()->getName() << "\n");
+
+  // Actually allocate each operation.
+  for (auto &[storage, ops] : opsByStorage)
+    allocateOps(storage, block, ops);
+}
+
+void AllocateStatePass::allocateOps(Value storage, Block *block,
+                                    ArrayRef<Operation *> ops) {
+  SmallVector<std::tuple<Value, Value, IntegerAttr>> gettersToCreate;
+
+  // Helper function to allocate storage aligned to its own size, or 8 bytes at
+  // most.
+  unsigned currentByte = 0;
+  auto allocBytes = [&](unsigned numBytes) {
+    if (numBytes >= 2)
+      currentByte = (currentByte + 1) / 2 * 2;
```

Comment:
Maybe llmvm::alignTo
```suggestion
    if (numBytes >= 2)
      currentByte = llvm::alignTo(currentByte, 2);
```

Reply:
Implemented in a separate PR.